csrc/inlining.cpp

```diff
@@ -131,29 +143,97 @@ size_t MaxPosCalculator::getMaxPosSelf(
 size_t MaxPosCalculator::getMaxProducerPosFromConsumer(
     TensorView* producer,
     TensorView* consumer,
-    bool best_effort) const {
-  auto pairwise_logical_map = PairwiseLogicalDomainMap(producer, consumer);
-  auto replay_CasP = BestEffortReplay::replayCasP(
-      consumer, producer, -1, pairwise_logical_map);
-  auto p2c_replay_map = replay_CasP.getReplay();
-
-  for (const auto producer_pos : c10::irange(producer->nDims())) {
-    // If the producer position is mismatching with the consumer, then we can
-    // not inline into this position, otherwise the max producer position of
-    // the consumer will become invalid and expression sort will fail.
-    if (TransformReplay::getMatchedLeafPosWithoutReplayCasP(
-            consumer, producer, producer_pos + 1) < 0) {
-      return producer_pos;
+    bool best_effort) {
+  // Here, we have two methods to analayze inlining positions. One is
+  // the legacy BestEffortReplay-based one that allow forwarding of
+  // broadcast domains. This is the default method when both the
+  // producer and consumer tensors have loop domains that are derived
+  // from logical domains.
+  //
+  // When either of the two tensors has non-conventional loop domains
+  // through TensorView::setLoopDomain, IdModel-based analysis is
+  // required. At this moment, the IdModel-based analysis does not
+  // implement the broadcast forwarding, and therefore inlining of
+  // broadcast-merged domains does not work with this approach. In
+  // fact, it is likely we don't implement the forwarding with this
+  // approach as it may not be necessary.
+  //
+  // At this moment, in order to keep the existing behavior as is and
+  // at the same time allow inlinig with explicitly set loop domains,
+  // the legacy method is used whenever both the producer and consumer
+  // have loop domains that are derived from their logical domains
+  // with no redundancy. Otherwise, the IdModel-based method is used.
+
+  // TODO: Consider caching these properties in TensorView as they
+  // could only change with setLoopDomain
+  const bool may_need_forwarding =
+      lower_utils::hasRootToLoopLinearTransformations(producer) &&
```

Comment:
Should we make this `ir_utils` instead of `lower_utils`? It's strange that inlining uses lower utils.